File: src/Serialization/HybridRow.Tests.Perf/RowReaderExtensions.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Tests.Perf
{
    using Microsoft.Azure.Cosmos.Core;
    using Microsoft.Azure.Cosmos.Core.Utf8;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.IO;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Layouts;

    internal static class RowReaderExtensions
    {
        public static Result VisitReader(this ref RowReader reader)
        {
            while (reader.Read())
            {
                Utf8Span path = reader.PathSpan;
                switch (reader.Type.LayoutCode)
                {
                    case LayoutCode.Null:
                    case LayoutCode.Boolean:
                    case LayoutCode.Int8:
                    case LayoutCode.Int16:
                    case LayoutCode.Int32:
                    case LayoutCode.Int64:
                    case LayoutCode.UInt8:
                    case LayoutCode.UInt16:
                    case LayoutCode.UInt32:
                    case LayoutCode.UInt64:
                    case LayoutCode.VarInt:
                    case LayoutCode.VarUInt:
                    case LayoutCode.Float32:
                    case LayoutCode.Float64:
                    case LayoutCode.Float128:
                    case LayoutCode.Decimal:
                    case LayoutCode.DateTime:
                    case LayoutCode.UnixDateTime:
                    case LayoutCode.Guid:
                    case LayoutCode.MongoDbObjectId:
                    case LayoutCode.Utf8:
                    case LayoutCode.Binary:
                        break;

                    case LayoutCode.NullableScope:
                    case LayoutCode.ImmutableNullableScope:
                    {
                        if (!reader.HasValue)
                        {
                            break;
                        }

                        goto case LayoutCode.TypedTupleScope;
                    }

                    case LayoutCode.ObjectScope:
                    case LayoutCode.ImmutableObjectScope:
                    case LayoutCode.Schema:
                    case LayoutCode.ImmutableSchema:
                    case LayoutCode.ArrayScope:
                    case LayoutCode.ImmutableArrayScope:
                    case LayoutCode.TypedArrayScope:
                    case LayoutCode.ImmutableTypedArrayScope:
                    case LayoutCode.TypedSetScope:
                    case LayoutCode.ImmutableTypedSetScope:
                    case LayoutCode.TypedMapScope:
                    case LayoutCode.ImmutableTypedMapScope:
                    case LayoutCode.TupleScope:
                    case LayoutCode.ImmutableTupleScope:
                    case LayoutCode.TypedTupleScope:
                    case LayoutCode.ImmutableTypedTupleScope:
                    case LayoutCode.TaggedScope:
                    case LayoutCode.ImmutableTaggedScope:
                    case LayoutCode.Tagged2Scope:
                    case LayoutCode.ImmutableTagged2Scope:
                    {
                        Result r = reader.ReadScope(null, (ref RowReader child, object _) => child.VisitReader());
                        if (r != Result.Success)
                        {
                            return r;
                        }

                        break;
                    }

                    default:
                    {
                        Contract.Assert(false, $"Unknown type will be ignored: {reader.Type.LayoutCode}");
                        break;
                    }
                }
            }

            return Result.Success;
        }
    }
}

File: experimental/java/exclusions/test/java/com/azure/data/cosmos/serialization/hybridrow/perf/RowReaderExtensions.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.perf;

import com.azure.data.cosmos.core.Reference;
import com.azure.data.cosmos.serialization.hybridrow.Result;

import static com.azure.data.cosmos.serialization.hybridrow.layouts.LayoutCode.IMMUTABLE_TAGGED2_SCOPE;

public final class RowReaderExtensions {
    public static Result VisitReader(Reference<RowReader> reader) {
        while (reader.get().Read()) {
            Utf8Span path = reader.get().getPathSpan();
            switch (reader.get().getType().LayoutCode) {
                case Null:
                case Boolean:
                case Int8:
                case Int16:
                case Int32:
                case Int64:
                case UInt8:
                case UInt16:
                case UInt32:
                case UInt64:
                case VarInt:
                case VarUInt:
                case Float32:
                case Float64:
                case Float128:
                case Decimal:
                case DateTime:
                case UnixDateTime:
                case Guid:
                case MongoDbObjectId:
                case Utf8:
                case Binary:
                    break;

                case NullableScope:
                case ImmutableNullableScope: {
                    if (!reader.get().getHasValue()) {
                        break;
                    }

                    // TODO: C# TO JAVA CONVERTER: There is no 'goto' in Java:
					goto case LayoutCode.TypedTupleScope
                }

                case ObjectScope:
                case ImmutableObjectScope:
                case Schema:
                case ImmutableSchema:
                case ArrayScope:
                case ImmutableArrayScope:
                case TypedArrayScope:
                case ImmutableTypedArrayScope:
                case TypedSetScope:
                case ImmutableTypedSetScope:
                case TypedMapScope:
                case ImmutableTypedMapScope:
                case TupleScope:
                case ImmutableTupleScope:
                case TypedTupleScope:
                case ImmutableTypedTupleScope:
                case TaggedScope:
                case ImmutableTaggedScope:
                case Tagged2Scope:
                case IMMUTABLE_TAGGED2_SCOPE: {
                    // TODO: C# TO JAVA CONVERTER: The following lambda contained an unresolved 'ref' keyword - these are not converted by C# to Java Converter:
                    Result r = reader.get().ReadScope(null, (ref RowReader child, Object _) -> child.VisitReader());
                    if (r != Result.SUCCESS) {
                        return r;
                    }

                    break;
                }

                default: {
                    throw new IllegalStateException(lenientFormat("Unknown type will be ignored: %s", reader.get().getType().LayoutCode));
                    break;
                }
            }
        }

        return Result.SUCCESS;
    }
}